traefik/templates/gateway.yaml

```diff
@@ -14,25 +17,42 @@ metadata:
 spec:
   gatewayClassName: {{ default "traefik" .Values.gatewayClass.name }}
   listeners:
-    - name: web
-      port: {{ .Values.ports.web.port }}
-      protocol: HTTP
-      {{- with .Values.gateway.namespacePolicy }}
-      allowedRoutes:
-        namespaces:
-          from: {{ . }}
+    {{- range $name, $config := .Values.gateway.listeners }}
+    - name: {{ $name }}
+      {{ $found := false }}
+      {{- range $portName, $portConfig := $.Values.ports -}}
+        {{- if eq $portConfig.port $config.port -}}
+          {{ $found = true }}
+        {{- end -}}
+      {{- end -}}
+      {{ if not $found }}
+        {{- fail (printf "ERROR: port %0.f is not declared in ports" .port ) }}
+      {{- end -}}
+      port: {{ .port }}
+      {{- if not (has .protocol (list "HTTP" "HTTPS" "TCP" "TLS" "UDP")) }}
```

Comment:
I'm wondering whether or not we should be doing this kind of validations. On one hand, additional validation is good, on the other this requires constant care on our side to match any upstream API changes. Although I do admit this one is unlikely to change and changing it requires little effort, I also wonder how much of a benefit this is in practice.